### PR TITLE
Generate AppImage

### PR DESCRIPTION
**Pull request by andyleejordan**
Replaces #2027 as it had merge conflicts due to macOS -> OS X rename. Rebased to fix failing Pester test run.

@probonopd and @TravisEz13 please review that I fixed conflicts correctly.

Push build is [here](https://travis-ci.org/andschwa/PowerShell/builds/155445453).

### CODE REVIEW

**Comment by andschwa** [comment]
Push build succeeded. Pester tests ran and AppImage was created successfully.

**Comment by probonopd** [comment]
Thanks, looks good to me.

**Comment by andschwa** [comment]
@TravisEz13 can you try out the AppImage?

**Comment by TravisEz13** [comment]
@andschwa I see the images generated here, but you don't have artifacts enabled so I cannot download them.  https://travis-ci.org/andschwa/PowerShell/builds/155445453

**Comment by andschwa** [comment]
@TravisEz13 ah, of course. Here's an upstream [build](https://travis-ci.org/PowerShell/PowerShell/builds/155773444) which should upload the artifacts when its complete.

**Comment by probonopd** [comment]
Upload succeeded, download_url: https://s3.amazonaws.com/powershell-artifacts/PowerShell/PowerShell/258/258.1/powershell-6.0.0-alpha.9-x86_64.AppImage

Will test now.

**Comment by probonopd** [comment]
Tested on Ubuntu 16.04, runs.

**Comment by andschwa** [comment]
LGTM. @TravisEz13?

**Comment by andschwa** [comment]
@probonopd I have a hold that came internally, we can't merge this yet.

@KrisBash can you please explain?

**Comment by andschwa** [comment]
@KrisBash nevermind, I spoke with @probonopd.

File: tools/appimage.sh
Comment (lzybkr) [inline tools/appimage.sh]: We shouldn't be fetching scripts from potentially untrusted sources as part of building our package - we should have copies of everything we need in our repo.

**Comment by andschwa** [comment]
Hm, I can't assign @KrisBash to the issue.

@lzybkr should we add him as a collaborator?

File: tools/appimage.sh
Comment (probonopd) [inline tools/appimage.sh]: functions.sh is part of the [probonopd/AppImages](https://github.com/probonopd/AppImages) project which is under the MIT License so you are free to mirror it but it is neither part of the PowerShell project nor subject to the Microsoft CLA. The binaries it downloads in the generate_appimage function are built from the sources in the the [probonopd/AppImageKit](https://github.com/probonopd/AppImageKit) project which is also under the MIT License but is also neither part of the PowerShell project nor subject to the Microsoft CLA. You are free to rebuild them from your own mirror, too, if you like.

**Comment by zaxebo1** [comment]
I am using powershell AppImage on linux , available at https://bintray.com/probono/AppImages/PowerShell .  It helps in single portable binary executable file getting executed on various linux distributions. 
**I will humbly request and very strongly recommend that each powershell release by the powershell development team, should also officially happen in AppImage format , and should be available at https://github.com/PowerShell/PowerShell/releases**

**Comment by andschwa** [comment]
Hello, and thank you for your input @zaxebo1. I am following up internally on why this is still blocked; we should have a better answer here soon. Thanks again :smile:

**Comment by SteveL-MSFT** [comment]
Working on getting resolution on this, won't happen til next week earliest

**Comment by andschwa** [comment]
Thanks for following up @SteveL-MSFT 😄

**Comment by SteveL-MSFT** [comment]
@andschwa this has been approved, but we must include some attribution, I'll send them to you via email to include in the image

**Comment by zaxebo1** [comment]
the above are **trivial conflicts** - that is, conflicts only in markdown files and a shell script, which can be quickly resolved. Request you to kindly take up  on early/priority - to resolve these conflicting files

**Comment by probonopd** [comment]
@zaxebo1 I think I don't have the permissions, only @andschwa can do this since he opened this PR based on a previous one from me.

**Comment by andschwa** [comment]
@probonopd is right, since I can fix this up next week.

**Comment by vors** [comment]
@andschwa ping

**Comment by andschwa** [comment]
@vors is there a !remindme bot for GitHub? If not, there should be. Adding this to my task list...

**Comment by vors** [comment]
I will play its role for now :)

**Comment by andschwa** [comment]
@vors this is ready now as far as I know.

**Comment by andschwa** [comment]
Steve can you send the attributions requirements again?

File: README.md
Comment (andschwa) [inline README.md]: Oops a typo.

**Comment by andschwa** [comment]
- [x] Rebased and merge conflicts resolved
- [x] Pinned external code download to specific commit hash
- [x] Included additional third party licenses from legal
- [x] Successful build of AppImage binary
- [ ] Anything else?

@PowerShell/powershell-maintainers you'll want to include the AppImage binary in the next release on GitHub, and fix the link in the readme.

**Comment by andschwa** [comment]
@SteveL-MSFT handing over to you for final review and merge 😄

**Comment by joeyaiello** [comment]
> Anything else?

@andschwa just *another* rebase ;)

**Comment by andschwa** [comment]
@joeyaiello your turn!

**Comment by probonopd** [comment]
Never ending story ;-) @andschwa

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Comment by probonopd** [comment]
Thanks for merging. Minor cosmetic glitch, in `README.md` it now says `[.AppImage][rl-ai]` which doesn't look right.

**Comment by SteveL-MSFT** [comment]
@raghushantha do you need to update your publishing pipeline to generate the appimage?  once we have an official appimage, can you update the link to point to it?

**Comment by andschwa** [comment]
Woo hoo!

**Comment by probonopd** [comment]
Let's celebrate @andschwa :1st_place_medal:

**Comment by raghushantha** [comment]
@SteveL-MSFT I have integrated AppImage generation as part of the Release pipeline.
Got a couple of concerns which I will followup offline

**Pull request merged**
